### PR TITLE
Better mergesort.py

### PR DESCRIPTION
**Pull request by LuisEnMarroquin**
Allow the function to be executed without having to pass the start and end indexes of the array

### **Describe your change:**

Added default values to `mergesort` function to automatically calculate the start and end indexes

Includes @cclauss revision

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?
* [x] Improved an algorithm?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: divide_and_conquer/mergesort.py [hunk 1]
```diff
@@ -27,13 +27,16 @@ def merge(arr, left, mid, right):
...
 
 
-def mergesort(arr, left, right):
+def mergesort(arr, left=None, right=None):
     """
-    >>> mergesort([3, 2, 1], 0, 2)
+    >>> mergesort([3, 2, 1])
     [1, 2, 3]
-    >>> mergesort([3, 2, 1, 0, 1, 2, 3, 5, 4], 0, 8)
+    >>> mergesort([3, 2, 1, 0, 1, 2, 3, 5, 4])
```
Comment (cclauss) [inline divide_and_conquer/mergesort.py:34]: Please add a doctest for `mergesort([3, 2, 1], -1, -3.1)`

Comment (LuisEnMarroquin) [reply divide_and_conquer/mergesort.py:34]: Added that doctest and also `mergesort([3, 2, 1], None, False)`

**Comment by shellhub** [comment]
@cclauss can you take a look why Travic CI build error. Thanks in advance.

**Comment by spamegg1** [comment]
> @cclauss can you take a look why Travic CI build error. Thanks in advance.

Build log says 
>E   _pytest.pathlib.ImportPathMismatchError: ('merge_sort', '/home/travis/build/TheAlgorithms/Python/divide_and_conquer/merge_sort.py', PosixPath('/home/travis/build/TheAlgorithms/Python/sorts/merge_sort.py'))

**Comment by cclauss** [comment]
Reverting the name change will fix the issue.  It is bugging me to have two different merge sorts in two different directories.  Maybe we should just have one in the sorts directory.

File: divide_and_conquer/mergesort.py [hunk 2]
```diff
@@ -27,13 +27,21 @@ def merge(arr, left, mid, right):
...
     [0, 1, 1, 2, 2, 3, 3, 4, 5]
+    >>> mergesort([3, 2, 1], -1, -3.1)
+    [1, 2, 3]
+    >>> mergesort([3, 2, 1], None, False)
+    [1, 2, 3]
     """
+    if left is None or left is False or left < 0:
```
Comment (cclauss) [inline divide_and_conquer/mergesort.py:41]:
```suggestion
    if (left or -1) < 0:
```

File: divide_and_conquer/mergesort.py [hunk 3]
```diff
@@ -27,13 +27,21 @@ def merge(arr, left, mid, right):
...
+        left = 0
+    if right is None or right is False or right < 0:
```
Comment (cclauss) [inline divide_and_conquer/mergesort.py:43]:
```suggestion
    if (right or -1) < 0:
```

**Comment by LuisEnMarroquin** [comment]
@cclauss That conditional won't work, value will always be -1

Try running this:

```python
def mergesort(num, left=None):
    yyy = (left or -1)
    if (left or -1) < 0:
        print('inside if', num, yyy)

mergesort(0)
mergesort(1, None)
mergesort(2, False)
mergesort(3, -1)
mergesort(4, 0)
mergesort(5, 1)
```

And you will get:

```shell
inside if 0 -1
inside if 1 -1
inside if 2 -1
inside if 3 -1
inside if 4 -1
```

**Comment by cclauss** [comment]
```python
>>> for i in range(-5, 5):
...     left = i
...     if (left or -1) < 0:
...         left = 0
...     print(i, left)
... 
-5 0
-4 0
-3 0
-2 0
-1 0
0 0
1 1
2 2
3 3
4 4
```

**Comment by LuisEnMarroquin** [comment]
> @cclauss That conditional won't work, value will always be -1
> 
> Try running this:
> 
> ```python
> def mergesort(num, left=None):
>     yyy = (left or -1)
>     if (left or -1) < 0:
>         print('inside if', num, yyy)
> 
> mergesort(0)
> mergesort(1, None)
> mergesort(2, False)
> mergesort(3, -1)
> mergesort(4, 0)
> mergesort(5, 1)
> ```
> 
> And you will get:
> 
> ```shell
> inside if 0 -1
> inside if 1 -1
> inside if 2 -1
> inside if 3 -1
> inside if 4 -1
> ```

Please run this, `or` operator is not working as expected inside **def**

Executed on:
* Ubuntu Server - Python 3.6.9
* Windows 10 Pro - Python 3.8.5

**Comment by LuisEnMarroquin** [comment]
After adding your suggested changes I get `maximum recursion depth exceeded`

image

Here is the file with your changes applied

```python
def merge(arr, left, mid, right):
    # overall array will divided into 2 array
    # left_arr contains the left portion of array from left to mid
    # right_arr contains the right portion of array from mid + 1 to right
    left_arr = arr[left : mid + 1]
    right_arr = arr[mid + 1 : right + 1]
    k = left
    i = 0
    j = 0
    while i < len(left_arr) and j < len(right_arr):
        # change sign for Descending order
        if left_arr[i] < right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1
    while i < len(left_arr):
        arr[k] = left_arr[i]
        i += 1
        k += 1
    while j < len(right_arr):
        arr[k] = right_arr[j]
        j += 1
        k += 1
    return arr


def mergesort(arr, left=None, right=None):
    """
    >>> mergesort([3, 2, 1])
    [1, 2, 3]
    >>> mergesort([3, 2, 1, 0, 1, 2, 3, 5, 4])
    [0, 1, 1, 2, 2, 3, 3, 4, 5]
    >>> mergesort([3, 2, 1], -1, -3.1)
    [1, 2, 3]
    >>> mergesort([3, 2, 1], None, False)
    [1, 2, 3]
    """
    if (left or -1) < 0:
        left = 0
    if (right or -1) < 0:
        right = len(arr) - 1
    if left < right:
        mid = (left + right) // 2
        # print("ms1",a,b,m)
        mergesort(arr, left, mid)
        # print("ms2",a,m+1,e)
        mergesort(arr, mid + 1, right)
        # print("m",a,b,m,e)
        merge(arr, left, mid, right)
        return arr


if __name__ == "__main__":
    import doctest

    doctest.testmod()

```

**Pull request closed without merging**